File: .tmp_c4000/batch_04_go/3489_zero_array_transformation_iv.go
// LeetCode 3489 - Zero Array Transformation IV
// https://leetcode.com/problems/zero-array-transformation-iv/

func minZeroArray(nums []int, queries [][]int) int {
	n := len(nums)
	ok := func(k int) bool {
		for i := 0; i < n; i++ {
			if nums[i] == 0 {
				continue
			}
			// can we form nums[i] using subset of query values covering i among first k?
			vals := []int{}
			for q := 0; q < k; q++ {
				l, r, v := queries[q][0], queries[q][1], queries[q][2]
				if l <= i && i <= r {
					vals = append(vals, v)
				}
			}
			if !canSubsetSum(vals, nums[i]) {
				return false
			}
		}
		return true
	}
	if ok(0) {
		return 0
	}
	lo, hi := 1, len(queries)+1
	for lo < hi {
		mid := (lo + hi) / 2
		if mid <= len(queries) && ok(mid) {
			hi = mid
		} else {
			lo = mid + 1
		}
	}
	if lo > len(queries) {
		return -1
	}
	return lo
}

func canSubsetSum(vals []int, target int) bool {
	if target == 0 {
		return true
	}
	dp := make([]bool, target+1)
	dp[0] = true
	for _, v := range vals {
		for s := target; s >= v; s-- {
			if dp[s-v] {
				dp[s] = true
			}
		}
	}
	return dp[target]
}

File: .tmp_rs4013/rust_check17/3489_zero_array_transformation_iv.rs
struct Solution;
// LeetCode 3489 - Zero Array Transformation IV
// https://leetcode.com/problems/zero-array-transformation-iv/

impl Solution {
    fn can_subset_sum(vals: &[i32], target: i32) -> bool {
        if target == 0 {
            return true;
        }
        let mut dp = vec![false; (target + 1) as usize];
        dp[0] = true;
        for &v in vals {
            for s in (v..=target).rev() {
                if dp[(s - v) as usize] {
                    dp[s as usize] = true;
                }
            }
        }
        dp[target as usize]
    }

    pub fn min_zero_array(nums: Vec<i32>, queries: Vec<Vec<i32>>) -> i32 {
        let n = nums.len();
        let ok = |k: i32| -> bool {
            for i in 0..n {
                if nums[i] == 0 {
                    continue;
                }
                let mut vals = Vec::new();
                for q in 0..k as usize {
                    let l = queries[q][0];
                    let r = queries[q][1];
                    let v = queries[q][2];
                    if l <= i as i32 && i as i32 <= r {
                        vals.push(v);
                    }
                }
                if !Self::can_subset_sum(&vals, nums[i]) {
                    return false;
                }
            }
            true
        };
        if ok(0) {
            return 0;
        }
        let mut lo = 1;
        let mut hi = queries.len() as i32 + 1;
        while lo < hi {
            let mid = (lo + hi) / 2;
            if mid <= queries.len() as i32 && ok(mid) {
                hi = mid;
            } else {
                lo = mid + 1;
            }
        }
        if lo > queries.len() as i32 {
            -1
        } else {
            lo
        }
    }
}

fn main() {}
